rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,145 @@
+- **Feature Name:** Loop-contract synthesis
+- **Feature Request Issue:** <https://github.com/model-checking/kani/issues/2214>
+- **RFC PR:** <https://github.com/model-checking/kani/pull/2215>
+- **Status:** Under Review
+- **Version:** 0
+- **Proof-of-concept:** <https://github.com/qinheping/kani/tree/kani-synthesizer>
+
+## Summary
+
+An new option that allows users to verify programs without unwinding loops by synthesizing loop contracts for those loops.
+
+
+## User Impact
+Currently Kani does not support proving correctness (i.e. assertions never fail) on programs with unbounded control flow (e.g. loops with dynamic bounds). Kani unrolls all unbounded loops until a few iterations and then verifies this unrolled program — it thus provides a much weaker guarantee on correctness.
+
+The loop-contract synthesizer will synthesize loop contracts for loops without unwinding numbers specified. With the synthesized loop contracts, CBMC abstracts those loops as non-loop structures that can cover arbitrary iterations of the loops. As a result, with the synthesizer proposed in this RFC, Kani can verify programs without unwind loops, and provide a stronger guarantee on correctness. For example, the number of executed iterations of the loop in the following harness is dynamically bounded by an unbounded variable `y`; hence no unwinding number can guarantee the correctness of the loop.
+```rust
+#[kani::proof]
+fn main() {
+    let mut y: i32 = kani::any_where(|i| *i > 0, "y is a positive integer");
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+With the loop-contract synthesizer, Kani can synthesize the loop invariant `y >= 0`, with which it can prove the post-condition `y == 0` without unwinding the loop.
+
+
+Also, these contracts, when provided, would also significantly improve Kani’s verification time since all loops would be unrolled only to a single iteration, as opposed to a number of iterations which is Kani’s current behavior. For example, although the following loop is bounded and we can easily find out the unwinding number `50000` for it, unwinding the loop 50000 times will lead to a large query, and hence slow down Kani. With loop contract, we only verify the single abstract iteration of the loop, which leads to a smaller query.
+```rust
+#[kani::proof]
+#[kani::unwind(50000)]
+fn main() {
+    let mut y: i32 = 50000;
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+
+
+## User Experience
+
+Once this RFC has been stabilized, users will be able to synthesize loop contracts for loops and verify programs with the synthesized loop contracts with a new command line flag `--synthesize-loop-contracts`.
```

Comment:
That's why it's important to specify for which loops will Kani support in this RFC.